python-sdk/nuscenes/eval/predict/metrics.py

```diff
@@ -0,0 +1,315 @@
+    :param metric_results: 1-dimensional array of shape [batch_size, num_modes]
+    :param mode_probabilities: 1-dimensional array of shape [batch_size, num_modes]
+    :param ranking_fcn: Either 'min' or 'max'. How you want to metrics ranked over the top
+            k modes.
+    :return: Array of shape [num_modes]
+    """
+
+    if ranking_func == "min":
+        func = np.minimum.accumulate
+    elif ranking_func == "max":
+        func = np.maximum.accumulate
+    else:
+        raise ValueError(f"Parameter ranking_func must be one of min or max. Received {ranking_func}")
+
+    p_sorted = np.flip(mode_probabilities.argsort(axis=-1), axis=-1)
+    indices = np.indices(metric_results.shape)
+
+    sorted_metrics = metric_results[indices[0], p_sorted]
+
+    return func(sorted_metrics, axis=-1)
+
+
+def hit_rate_top_k(stacked_trajs: np.ndarray, stacked_ground_truth: np.ndarray,
+                   mode_probabilities: np.ndarray,
+                   tolerance: float) -> np.ndarray:
+    """Compute the hit rate over the top k modes."""
+
+    hit_rate = hit_max_distances(stacked_trajs, stacked_ground_truth, tolerance)
+    return rank_metric_over_top_k_modes(hit_rate, mode_probabilities, "max")
+
+
+def min_ade_k(stacked_trajs: np.ndarray, stacked_ground_truth: np.ndarray,
+              mode_probabilities: np.ndarray) -> np.ndarray:
+    """Compute the min ade over the top k modes."""
+
+    ade = mean_distances(stacked_trajs, stacked_ground_truth)
+    return rank_metric_over_top_k_modes(ade, mode_probabilities, "min")
+
+
+def min_fde_k(stacked_trajs: np.ndarray, stacked_ground_truth: np.ndarray,
+              mode_probabilities: np.ndarray) -> np.ndarray:
+    """Compute the min fde over the top k modes."""
+
+    fde = final_distances(stacked_trajs, stacked_ground_truth)
+    return rank_metric_over_top_k_modes(fde, mode_probabilities, "min")
+
+
+def stack_ground_truth(ground_truth: np.ndarray, num_modes: int) -> np.ndarray:
+    """Make k identical copies of the ground truth to make computing the metrics across modes
+    easier.
+    :param ground_truth: shape [horizon_length, state_dim]
+    :param num_modes: number of modes in prediction
+    :return: shape [num_modes, horizon_length, state_dim]
+    """
+    return np.repeat(np.expand_dims(ground_truth, 0), num_modes, axis=0)
+
+
+class SerializableFunction(abc.ABC):
+    """Function that can be serialized/deserialized to/from json."""
+
+    @abc.abstractmethod
+    def serialize(self) -> Dict[str, Any]:
+        pass
+
+    @property
+    @abc.abstractmethod
+    def name(self,) -> str:
+        pass
+
+
+class Aggregator(SerializableFunction):
+    """Function that can aggregate many metrics across predictions."""
+
+    @abc.abstractmethod
+    def __call__(self, array: np.ndarray, **kwargs) -> List[float]:
+        pass
+
+
+class RowMean(Aggregator):
+
+    def __call__(self, array: np.ndarray) -> np.ndarray:
```

Comment:
`Aggregator` has another *kwargs argument. I've added it here so that both have the same signature.